src/BinomialQueue/Internals.hs

```diff
@@ -331,9 +333,21 @@ incrExtract :: Extract (Succ rk) a -> Extract rk a
 incrExtract (Extract minKey (Succ kChild kChildren) ts)
   = Extract minKey kChildren (Cons kChild ts)
 
+-- Note: We used to apply Skip lazily here, for fear that the incr application
+-- might cause a cascade of thunks to be forced. However, that is not actually
+-- possible. If ts starts with a Skip, then we simply replace it without
+-- forcing its child. The only way the forest in an 'Extract' can start with a
+-- Cons is if it was produced by 'incrExtract'. But in that case, its next node
+-- came from an 'Extract', and is therefore already in WHNF. So if the 'incr'
```

Comment:
How can we see that the next node is already in WHNF? Is the spine of the `BinomForest` in `Extract` always forced?

Reply:
`incrEctract` gets a node out of `Extract` (which must be in WHNF because `Extract` is a strict constructor) and applies `Cons` to it.

Reply:
And `incrExtract` is the only function that might put a `Cons` in an `Extract`?

Reply:
That's what it looks like to me. I certainly wouldn't mind if you double checked. However, as long as you're running benchmarks, could you see what happens if we embrace the cascade and switch `incr` to `incr'` here?

Reply:
It looks like that to me as well. Let's add that to the comment.

Using `incr'` has more mixed results (results vary a bit between runs), but overall seems to further improve things (10-20% less than just using `$!`). Wouldn't that increase the runtime complexity of `incrExtract'` from O(1) to O(n) though?

Reply:
No, it would increase it from $O(1)$ to $O(\log n)$. But the sum total of all the `incrExtract'`s would remain $O(\log n)$, because of the way the carries work out. I'm pretty sure the worst case scenario is when the `BinomHeap` is one more than a power of two and we delete from the big tree. In that case, we'll spread the big tree all the way across the spine, then propagate carries all the way back. In other situations, we'll hit one or more `Cons`es along the way, which will lead to propagating in multiple segments, more likely to be in L1.

Reply:
Should I just switch to `incr'`? I'm fine with that.

Reply:
If you think that wouldn't change `incrExtract'`s complexity, that's fine by me.

Reply:
I'm confident it's fine for big $O$. Each cascade leaves a `Skip` at its head which will interrupt the next (shallower) cascade. Okasaki explains this particular thing somewhere or other in his book.